ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/Tests.scala

```diff
@@ -11,6 +11,7 @@ object Tests {
 
   val default: Tests = Map(
     "SemanticTests" -> (new SemanticTests(_)),
+    "PartyAllocationIT" -> (new PartyAllocation(_)),
```

Comment:
So far we only kept `SemanticTests` in the default run, everything else is opt-in. Any reason to change this?

Reply:
Checking for party allocation in a transaction looks like a kind of semantic (well-formedness) test and ideally I think we'd like to check it for anything that is not the sandbox (until it is migrated to KVUtils and pre-allocates parties) and Canton (it already fails but with a different message).

Reply:
> `CHANGELOG_BEGIN`
> [DAML Ledger Integration Kit] Enforce that all parties referenced in a transaction have been allocated
> `CHANGELOG_END`

I suggest to change this to the text
```
[DAML Ledger Integration Kit] Enforce that all parties referenced as stakeholders, actors, or maintainers in a transaction have been allocated
```

That means we should *not* check party literals referenced in DAML values in transaction nodes. This is in line with the documentation of [partyFromText](https://docs.daml.com/daml/reference/base.html#function-da-internal-lf-partyfromtext-40570) which explicitly states that we do not check whether the party exists on the ledger when converting from `Text` to `Party`. For example this could be used in an update like this to bring into scope a party value during evaluation that is not referenced in the transaction itself.
``` 
do let Some p1 = partyFromText "Alice"
    let Some p2 = partyFromText "Bob"
    assert (p1 /= p2)
```
(Better examples should be constructible ;-) )

The design decision for `partyFromText` was made to make it a pure function.

Apart from the above reason there is also a deeper technical reason: we would like to keep the option to introduce generic templates, which are parametric in the type that they instantiate. So the behavior of `template GenFoo a = ...` should not be different between `instantiate template GenFoo Int` and `instantiate template GenFoo Party`. If we start peeking into the DAML values of a transaction at commit time this does no longer hold.

@fabiotudone-da : would it be possible to change this PR to only forbid unallocated actors, mantainers, and stakeholders? What problems/arguments am I overlooking?

@martin-drhu-da : can you double-check my reasoning?

Reply:
This PR originates from a bug found by @mziolekda: if Bob is not allocated and Alice creates a contract with Bob as parameter, where Bob is a controller in the template body but not a signatory nor an observer, the transaction goes through, an implicit Bob party is referenced and any subsequent attempt to allocate a Bob party fails (i.e. even without Bob exercising any choice).

We assumed that the implicit party reference is unavoidable (which maybe isn't) and concentrated on whether allowing unallocated parties appearing anywhere in transactions (incl. values) make sense at all.

We discussed the topic and came to the conclusion that unallocated parties:

- Either aren't used at all, or they will need to be allocated before being stakeholders, or transactions involving them won't go through. Given that fact, from an application logic point of view, they should really be allocated before they are referenced or they should be strings until it makes sense for them to become parties (after being allocated). In this sense, enforcing that they are all allocated no matter where they appear might detect logic errors earlier.
- Assuming we should reject "create" actions where a controller expression evaluates to an unallocated party (but maybe we shouldn't and we should rather look into those "phantom" parties), we should enrich the "create" action model to include controller party values.
- From a purely conceptual point of view, unallocated parties feel "dangling" just like a contract ID not referencing any contract.

@dajmaki and @gerolf-da if I remember correctly you also raised issues related to history verification but I can't remember them precisely right now.

As for generic templates, (type) instantiation would not behave differently, I think, but a creation action of a `GenFoo Party` contract might fail for the additional reason that the party is not allocated (if the type parameter is used in the `with` clause).

On the other hand it is certainly true that traversing values is an additional verification dimension with a computational cost attached.

Reply:
Thanks for the details @fabiotudone-da . I'm not sure that the bug you mention (quoted below) works exactly as described.

> This PR originates from a bug found by @mziolekda: if Bob is not allocated and Alice creates a contract with Bob as parameter, where Bob is a controller in the template body but not a signatory nor an observer, the transaction goes through, an implicit Bob party is referenced and any subsequent attempt to allocate a Bob party fails (i.e. even without Bob exercising any choice).

Points of invstigation: 
1. I would like to see the exact example of how we have a controller that is not an observer. The usual desuguaring of `controller x can` is that `x` is added to the observers.  
2. I would like to understand why creating an instance `Foo 'Alice' 'Bob'` of a contract `template Foo with s : Party; v : Party where signatory s` is blocking from allocating the party 'Bob'.

Reply:
1. Yes you're right, that was me getting confused, Bob is an observer there.
2. As pointed out by @dajmaki this happens locally in the indexer (which is also shared by the sandbox) and was probably a left-over from the "open-world mode" elimination some months ago.

Trying to summarize the way forward we agreed upon, for alignment and future reference:

1. "Closed-world mode" is the correct semantics and should be tested in the test tool for all ledgers, including the sandbox (TBD how to reach this end state but implicit allocation, if at all needed, should probably pushed up to some "experience simplification" layer).
2. Parties allocation validation should be expanded to include all "informees" but not purely "named" parties (i.e. just mentioned in values)